runtime/storage.go

```diff
@@ -379,81 +379,95 @@ func (s *Storage) commit(inter *interpreter.Interpreter, commitContractUpdates b
-//	getDomainStorageMap := func(
-//		ledger atree.Ledger,
-//		storage atree.SlabStorage,
-//		address common.Address,
-//		domain common.StorageDomain,
-//	) (*interpreter.DomainStorageMap, error) {
-//		domainStorageKey := interpreter.NewStorageDomainKey(s.memoryGauge, address, domain)
-//
-//		// Get cached domain storage map if available.
-//		domainStorageMap := s.cachedDomainStorageMaps[domainStorageKey]
-//
-//		if domainStorageMap != nil {
-//			return domainStorageMap, nil
-//		}
-//
-//		return getDomainStorageMapFromV1DomainRegister(ledger, storage, address, domain)
-//	}
-//
-//	migrator := NewDomainRegisterMigration(s.Ledger, s.PersistentSlabStorage, inter, s.memoryGauge)
-//	migrator.SetGetDomainStorageMapFunc(getDomainStorageMap)
-//
-//	migratedAccounts, err := migrator.MigrateAccounts(s.unmigratedAccounts, migrateAccountPred)
-//	if err != nil {
-//		return err
-//	}
-//
-//	if migratedAccounts == nil {
-//		return nil
-//	}
-//
-//	// Update internal state with migrated accounts
-//	for pair := migratedAccounts.Oldest(); pair != nil; pair = pair.Next() {
-//		address := pair.Key
-//		accountStorageMap := pair.Value
-//
-//		// Cache migrated account storage map
-//		accountStorageKey := interpreter.NewStorageKey(s.memoryGauge, address, AccountStorageKey)
-//		s.cachedAccountStorageMaps[accountStorageKey] = accountStorageMap
-//
-//		// Remove migrated accounts from unmigratedAccounts
-//		s.unmigratedAccounts.Delete(address)
-//	}
-//
-//	return nil
-//}
+func (s *Storage) migrateV1AccountsToV2(inter *interpreter.Interpreter) error {
+
+	if s.cachedV1Accounts == nil {
+		return nil
+	}
+
+	// getDomainStorageMap function returns cached domain storage map if it is available
+	// before loading domain storage map from storage.
+	// This is necessary to migrate uncommitted (new) but cached domain storage map.
+	getDomainStorageMap := func(
+		ledger atree.Ledger,
+		storage atree.SlabStorage,
+		address common.Address,
+		domain common.StorageDomain,
+	) (*interpreter.DomainStorageMap, error) {
+		domainStorageKey := interpreter.NewStorageDomainKey(s.memoryGauge, address, domain)
+
+		// Get cached domain storage map if available.
+		domainStorageMap := s.cachedDomainStorageMaps[domainStorageKey]
+
+		if domainStorageMap != nil {
+			return domainStorageMap, nil
+		}
+
+		return getDomainStorageMapFromV1DomainRegister(ledger, storage, address, domain)
+	}
+
+	migrator := NewDomainRegisterMigration(
+		s.Ledger,
+		s.PersistentSlabStorage,
+		inter,
+		s.memoryGauge,
+		getDomainStorageMap,
+	)
+
+	for pair := s.cachedV1Accounts.Oldest(); pair != nil; pair = pair.Next() {
+		address := pair.Key
+		isV1 := pair.Value
+
+		if !isV1 || !s.PersistentSlabStorage.HasUnsavedChanges(atree.Address(address)) {
+			continue
+		}
+
+		accountStorageMap, err := migrator.MigrateAccount(address)
+		if err != nil {
+			return err
+		}
+
+		// TODO: is this all that is needed?
```

Comment:
The update of the bookkeeping is the same as before: Make V2 aware of the new account storage map, so it is not re-loaded, and consider the account no longer as V1. From what I can see this should be sufficient, but it's good to double check this is sufficient. @fxamacker